app/src/main/java/com/example/mvpmusicapp/ui/MainActivity.kt

```diff
@@ -0,0 +1,248 @@
+        songAdapter = SongAdapter(this)
+        songAdapter!!.setOnClickSongItem(object : OnClickSongItem {
+            override fun onClickSong(position: Int) {
+                currentSong = position
+                musicService?.switchMusic(currentSong)
+                musicService?.playMusic()
+                startService()
+                setConstraintMusicDialog()
+                setProgressBar()
+                updateProgress()
+            }
+        })
+    }
+
+    private fun setProgressBar() {
+        binding.progressMusic.max = musicService?.getDuration()!!
+        binding.progressMusic.progress = 0
+    }
+
+    private fun initData() {
+        songPresenter = SongPresenter(
+            SongRepository(QueryLocalData(this)), this
+        )
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            when (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
+                PackageManager.PERMISSION_GRANTED -> songPresenter!!.getLocalSong()
+                else -> requestPermissions(
+                    arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
+                    storageRequest
+                )
+            }
+        }
+        bindService(musicIntent, serviceConnection, Context.BIND_AUTO_CREATE)
+        registerReceiver(musicReceiver, IntentFilter(getString(R.string.intent_action)))
+    }
+
+    private fun setConstraintMusicDialog() {
+        binding.apply {
+            cardPlayMusic.visibility = View.VISIBLE
+            imagePlayMusic.setImageResource(R.drawable.ic_pause)
+            updateMusicName()
+        }
+
+        val constraintSet = ConstraintSet()
+        constraintSet.clone(binding.constraintMainParent)
+        constraintSet.connect(
+            R.id.recycler_song,
+            ConstraintSet.BOTTOM,
+            R.id.card_play_music,
+            ConstraintSet.TOP
+        )
+        constraintSet.applyTo(binding.constraintMainParent)
+    }
+
+    override fun getAllSongList(songList: List<Song>) {
+        this.songList = songList.toMutableList()
+        songAdapter!!.setSongList(this.songList)
+        binding.recyclerSong.adapter = songAdapter
+        musicService?.setSongList(this.songList)
+    }
+
+    override fun showErrorMessage(message: String) {
+        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
+    }
+
+    override fun setPlayButton() {
+        musicService?.pauseMusic()
+        binding.imagePlayMusic.setImageResource(R.drawable.ic_play_button)
+    }
+
+    override fun setPauseButton() {
+        musicService?.resumeMusic()
+        binding.imagePlayMusic.setImageResource(R.drawable.ic_pause)
+    }
+
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<out String>,
+        grantResults: IntArray
+    ) {
+        if (requestCode == storageRequest && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+            songPresenter!!.getLocalSong()
+        }
```

Comment:
có quá nhiều assert non-null, c kiểm tra lại nhé

Reply:
![image](https://user-images.githubusercontent.com/53604984/117103311-4e9bb080-ada4-11eb-94dd-6045e6a5a774.png)
T sẽ sửa lại nhưng có 1 vài chỗ dùng toán tử thì nó bắt pải assert non null đó